[AIChat]: Add URL based routing for different chats

fallaciousreasoning opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/42055

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

petemill commented on components/ai_chat/resources/page/components/header/index.tsx:
```diff
@@ -51,7 +49,7 @@ export const PageTitleHeader = React.forwardRef(function (props: FeatureButtonMe
             kind='plain-faint'
             fab
-            onClick={() => { aiChatContext.onSelectConversationUuid(undefined) }}
+            onClick={() => location.href = "/"}
```
It doesn't seem like we're differentiating anymore between "start a new conversation" and "go to the last conversation for the active tab". Am I wrong about that? I'm doing a build now to verify.

fallaciousreasoning replied:
New conversation is `/`, once you've started a conversation, the URL will update to include the `uuid`

fallaciousreasoning replied:
Fixed this!

petemill reviewed (changes requested):
I think I'm spotting a couple of issues with the refactor to support this (essentially removing the logic from `chat_ui.tsx`):

1. Sidebar is showing the header for all conversations, when it should only when a conversation is chosen which isn't the most recent default for that tab, i.e. `undefined` conversation was asked for.
2. Navigating away from the active tab is no longer causing a new conversation to be shown.
3. ConversationContext is no longer independent of its embedding. The reason I had tried to keep all page-related logic in the page-specific UI (`chat_ui.tsx`) is so that ConversationContext can be moderately portable for the purposes of moving / duplicating it to an iframe, (or even showing a conversation in a popup in the future, but that's imagination right now). I had also in the past gone back and forth between it being provided the mojom bindings or it creating its own mojom bindings, but ended up with it being provided them so that the complicated logic for binding can be at a higher level and separated.

I think the URL pattern is good, I like that the page is `/` until there's conversation history, and then changes to `/{ID}`. However, navigating either to `/` or to `/{ID}` should result in _either_ a new conversation or a default conversation for the tab, depending on what the action was to cause that navigation:

- Initial Load: get the default conversation for the active tab `API.bindConversation(undefined)`
- Clicking back button: get the default conversation for the active tab
- Clicking "New conversation" button: start a new conversation BUT also mark that we haven't selected a specific conversation ID, so that subsequent navigations change the conversation displayed - `API.newConversation(); setSelectedConversationUuid(undefined);`
- UIObserver.onNewDefaultConversation: get the default conversation only if we're already showing a default conversation - `API.bindConversation(undefined)`

The difference comes down to the user asking for a specifc conversation vs. using the default context-related conversation and having it change with navigations. And it is all sidebar-only problems, except they share binding logic so whilst we might be able to restrict the URL navigation handling to the standalone page, the binding will be shared, and mobile will need the sidebar-functionality but with the URLs of the fullpage.

I like how, even though you've removed some of the explicitness with passing props to ConversationContext, that you've made a layer dealing with the URL handling (`useSelectedConversation`) but it's possibly not exclusive enough. There might be a danger in trying to make a common `NavigationContext` helper that we're overcomplicating a (fairly) simple API and making a mini-framework instead of keeping it application-specific and more flexible with more context (context not `Context` 😝!).

I suppose I'm just asking if we can keep some of the logic the way it is to solve removing the above features _and_ make it more readable (which is subjective if I'm the one that wrote the code and the only one that knew why it did it that way!). I'm sure there's a better way to handle the default conversation vs specific conversation logic _and_ support framing _and_ support sidebar/ fullpage / mobile (without dealing with time tokens!) but it works and the logic is in one place.

fallaciousreasoning commented:
@petemill I think I've fixed 1. and 2. by special casing the `default` chatId. I think this makes it slightly more obvious that it gets treated differently (I'd be tempted to make it `/tab` or something even more specific).

wrt 3, the binding logic is still in a separate component, its just inside the same file now. I could lift it into a different component, if you prefer that.

I've added another commit so it doesn't depend on `useSelectedConversation` anymore :smile:

petemill commented:
Yes probably a different file is neccessary because ConversationContext will be inside an iframe and AIChatContext will be on the outside (and AIChatContext and ConversationContext might each be split in to 2 for a reduced command set for the frame).

I'm coming around to `/default` (or something similar to `/associated-content`), but I did like that it would change to the conversation ID when the conversation was committed. I would say it's unimportant because it's Side Panel and noone will see the URL, except I think mobile is going to use the Side Panel pattern but in a real Tab, so the URL will be visible but not copyable / bookmarkable - not a permalink.

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-navigation/storybook.html)

diracdeltas commented:
depending on the answer to https://github.com/brave/brave-browser/issues/42055#issuecomment-2455617382 this may require security review

petemill commented on components/common/navigation/Context.tsx:
```diff
@@ -0,0 +1,135 @@
+export type NavigationParams = { [key: string]: string | undefined }
+export type Routes = Map<string, Array<undefined | ((params: NavigationParams) => void)>>;
+
+export interface NavigationContext {
```
Especially because this is a shared module, this could benefit from some comments / documentation at least at the top

petemill commented on components/ai_chat/resources/page/routes.tsx:
```diff
@@ -0,0 +1,33 @@
+    '/{chatId}'
+]
+
+export function Routes() {
```
It's not clear to me why this is a react component - could it be just a hook (since all it does it call hooks, i.e. `useAIChatRoutes()`?

petemill commented on components/ai_chat/resources/page/state/conversation_context.tsx:
```diff
@@ -7,16 +7,19 @@ import * as React from 'react'
 export interface ConversationContextProps {
   conversationHandler: API.ConversationHandlerRemote
   callbackRouter: API.ConversationUICallbackRouter
+  selectedConversationId: string | undefined
```
I think ideally we'd keep this out of the ConversationContext and both it and the knowledge of URLs / selected conversation out of this file. I can imagine we're not far off having conversations in other UIs (e.g. popups) outside of the AI Chat full-page / sidebar. We're probably more farther off having more than 1 conversation displayed at a time, but it's possible... All that to say it seems like URL / selected conversation should be handled at a higher, "page", level rather than a component responsible for only part of the page state/UI. Whether it's AIChatContext or something new like AIChatPageContext 🤷 
What do you think?

fallaciousreasoning replied:
Sweet, added a new ActiveConversationProvider :smile:

petemill commented on components/ai_chat/resources/page/components/header/index.tsx:
```diff
@@ -35,22 +36,20 @@ export const ConversationHeader = React.forwardRef(function (props: FeatureButto
             kind='plain-faint'
             fab
-            onClick={() => { aiChatContext.onSelectConversationUuid(undefined) }}
+            onClick={() => location.href = "/default"}
```
If we do have a sensible context responsible for url management, perhaps we should not have to repeat the url paths, and even make functions, e.g. `AIChatRouteContext.urlForTabConversation` 🤷

fallaciousreasoning replied:
I just exported a string for now.

petemill commented on components/ai_chat/resources/page/state/conversation_context.tsx:
```diff
@@ -498,6 +508,57 @@ export function ConversationContextProvider(
+
+    // The default conversation changes as the associated tab navigates, so
+    // listen for changes.
+    if (selectedConversation === 'default') {
```
How about we go with your suggestion of `'tab'`?

fallaciousreasoning commented:
@petemill I've refactored the code to remove the `NavigationContext` in favor of a simpler Hook, which doesn't require us to register routes upfront.

Additionally, I've lifted the `selectedConversation` logic into `ActiveConversationProvider` which handles creating/updating `conversationAPIs` in response to changes in the selected id. 

I've provided a default implementation of this interface which gets the selected conversation from the current URL. All URL related logic has been moved to this class.

petemill commented on components/ai_chat/resources/page/state/ai_chat_active_chat_provider.tsx:
```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
```
Do you think this file needs the `ai_chat_` prefix? Its export doesn't have that prefix and we're already in the `ai_chat` component. I think the ai_chat_context.tsx` has that prefix since it's meant to be the root context.

petemill commented on components/ai_chat/resources/page/state/ai_chat_active_chat_provider.tsx:
```diff
@@ -0,0 +1,104 @@
+  }, [visibleConversations, selectedConversationId])
+
+  return conversationAPI
+    ? children(details as any)
```
What's the reason for this slightly non-standard way of passing the data? Why not just use context? Is it to protect it from descendent components and make sure only the direct child gets access to the data? Either way, I think we need a comment at the top of this explaining its use.

fallaciousreasoning replied:
This is the render props pattern which is pretty standard in React - basically, setting up another context seemed like overkill here, especially because only the `ConversationContext` is concerned with the data it provides.

fallaciousreasoning replied:
https://legacy.reactjs.org/docs/render-props.html

fallaciousreasoning replied:
(added the comment)

petemill replied:
ok I think I see why you prefer this over a custom hook, just about...

petemill reviewed (changes requested):
When opening the sidebar to the tab-default conversation, I'm getting the conversation title header showing, instead of the default sidebar header. This conversation title header should only show when picking a conversation from the conversation list.
The default sidebar header should show:
- When opening the sidebar
- When clicking "new conversation" in the sidebar
- When clicking the back button from an archive conversation

petemill commented on components/ai_chat/resources/page/state/active_chat_provider.tsx:
```diff
@@ -0,0 +1,118 @@
+  updateSelectedConversationId: (conversationId: string | undefined) => void
+  conversationHandler: API.ConversationHandlerRemote
+  callbackRouter: API.ConversationUICallbackRouter
+  newConversation: () => void
```
semi-nit: this doesn't sound enough like an action. When I see it in the UI code, it can be confused for a newConversation variable. Suggestions: onNewConversation, handleNewConversation, createNewConversation

petemill commented on components/ai_chat/resources/page/state/conversation_context.tsx:
```diff
@@ -310,8 +307,20 @@ export function ConversationContextProvider(
+  }, [isVisible, updateSelectedConversationId])
+
+  React.useEffect(() => {
+    setPartialContext({ isTabAssociated: selectedConversationId === tabAssociatedChatId })
```
Would having `isTabAssociated` be provided from `ActiveChatContext` avoid an extra round of state setting? Does it matter?

fallaciousreasoning replied:
Ah yup, that was a hangover from before the ActiveChatContext was a context :smile:

petemill commented on components/ai_chat/resources/page/state/active_chat_provider.tsx:
```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
```
Should this file be named `ActiveChatContext` to be consistent and be obvious it contains a React Context, Context Provider and context state type?

brave-builds commented:
Released in [v1.75.52](https://github.com/brave/brave-browser/releases/tag/v1.75.52)
